src/embedded_flash/syscall.rs

```diff
@@ -157,83 +192,88 @@ impl Storage for SyscallStorage {
             return Err(StorageError::NotAligned);
         }
         let ptr = self.read_slice(index, value.len())?.as_ptr() as usize;
-
-        let code = unsafe {
-            syscalls::raw::allow(
-                DRIVER_NUMBER,
-                allow_nr::WRITE_SLICE,
-                // We rely on the driver not writing to the slice. This should use read-only allow
-                // when available. See https://github.com/tock/tock/issues/1274.
-                value.as_ptr() as *mut u8,
-                value.len(),
-            )
-        };
-        if code < 0 {
-            return Err(StorageError::CustomError);
-        }
-
-        let code = syscalls::command(DRIVER_NUMBER, command_nr::WRITE_SLICE, ptr, value.len());
-        if code.is_err() {
-            return Err(StorageError::CustomError);
-        }
-
-        Ok(())
+        write_slice_op(ptr, value)
     }
 
     fn erase_page(&mut self, page: usize) -> StorageResult<()> {
         let index = StorageIndex { page, byte: 0 };
         let length = self.page_size();
         let ptr = self.read_slice(index, length)?.as_ptr() as usize;
-        let code = syscalls::command(DRIVER_NUMBER, command_nr::ERASE_PAGE, ptr, length);
-        if code.is_err() {
-            return Err(StorageError::CustomError);
-        }
-        Ok(())
+        erase_page_op(ptr, length)
     }
 }
 
-fn find_slice<'a>(
-    slices: &'a [&'a [u8]],
-    mut start: usize,
-    length: usize,
-) -> StorageResult<&'a [u8]> {
-    for slice in slices {
-        if start >= slice.len() {
-            start -= slice.len();
-            continue;
+pub struct UpgradeLocations {
+    page_size: usize,
+    partitions: Vec<RangeInclusive<usize>>,
```

Comment:
It looks to me that this could simply be:
```rust
pub struct UpgradeLocations {
    page_size: usize,
    partition_pages: Vec<PageRange>,
    metadata_address: usize,
}
struct PageRange {
    address: usize,
    num_pages: usize,
}
```

Reply:
I thought more about what I actually want to support, including our team discussion. With the new higher-level API, I made both `partition` and `metadata` one range. I think we can assume that a firmware image will be consecutive, so it shouldn't be a problem for `partition`. And I wanted to keep `metadata` extensible, so it is a range too. This simplifies the implementation a bit, too.